### PR TITLE
Fix Flashinfer CUTLASS MOE Allgather

### PR DESCRIPTION
**Pull request by wenscarl**
# Fix the mismatch of size and input tensor shape of all-gatherv and reduce-scatterv.

## Purpose
In the DP + EP setup with chunking enabled, collective operations like allgatherv and reduce_scatterv require each participating rank to provide a list of local token sizes (from CPU). It is essential that all ranks share the same and correct local_sizes to maintain proper synchronization.

For example, with VLLM_MOE_DP_CHUNK_SIZE=256, suppose the initial (pre-chunking) local token sizes across three ranks are: [500, 200, 256].

To chunk these in a lockstep manner—ensuring that each rank contributes at least one token per iteration—the sizes are split as follows:

Iteration 1: 256, 200, 256
Iteration 2: 244,   1,   1
This results in 2 chunking iterations.

Prior to this PR, the logic for computing local_sizes was incorrect, which could lead to mismatched sizes and synchronization issues across ranks.

## Test Plan
```
# server.sh
VLLM_MOE_DP_CHUNK_SIZE=128 \
VLLM_WORKER_MULTIPROC_METHOD=0 CUDA_VISIBLE_DEVICES=0,1,2,3 VLLM_USE_STANDALONE_COMPILE=0 VLLM_USE_FLASHINFER_MOE_FP4=1 VLLM_WORKER_MULTIPROC_METHOD=spawn \
  vllm serve nvidia/DeepSeek-R1-FP4 \
    --quantization="modelopt_fp4" \
    --trust-remote-code \
    --max-model-len=2048 \
    --block-size=128 \
    --max-num-seqs=256 \
    --enable-expert-parallel \
    --load_format="dummy" \
    --gpu_memory_utilization=0.92 \
    --tensor-parallel-size 1 \
    --data-parallel-size 4 \
    --enforce-eager \
    --disable-log-requests > svr.log 2>&1 &
    
 # client.sh
VLLM_USE_FLASHINFER_MOE_FP4=1 VLLM_WORKER_MULTIPROC_METHOD=spawn \
python benchmarks/benchmark_serving.py \
  --model nvidia/DeepSeek-R1-FP4 \
  --dataset-name random \
  --ignore-eos \
  --num-prompts 512 \
  --max-concurrency 256 \
  --random-input-len 128 \
  --random-output-len 1024 
```

## Test Result
Start server.sh as above then client.sh. What to verify is to check the client.sh script pass.

### CODE REVIEW

File: vllm/distributed/device_communicators/cuda_communicator.py [hunk 1]
```diff
@@ -236,6 +236,8 @@ def _all_gather_single(input_: torch.Tensor,
             input_size = input_.size()
             if sizes is not None:
                 assert len(sizes) == world_size
+                if input_.shape[dim] != sizes[self.rank_in_group]:
+                    print(f"lhs:{input_.shape[dim]} vs rhs:{sizes[self.rank_in_group]}")
```
[the remark this answers is not in the document]
Comment (varun-sundar-rabindranath) [reply vllm/distributed/device_communicators/cuda_communicator.py]: maybe turn the assert below to, 
```
                assert input_.shape[dim] == sizes[self.rank_in_group], (
       f"lhs:{input_.shape[dim]} vs rhs:{sizes[self.rank_in_group]}")

```

File: vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py [hunk 2]
```diff
@@ -58,6 +41,18 @@ def topk_indices_dtype(self) -> Optional[torch.dtype]:
...
         return self.num_dispatchers_
 
+    @property
+    def num_tokens_across_dp_current(self):
+        return self._num_tokens_across_dp_current
+
+    def update_local_tokens(self, local_tokens: int):
+        if local_tokens is None:
+            raise ValueError("Local token count not set.")
+        self._num_tokens_across_dp_current = DPMetadata.num_tokens_across_dp(
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py]: @wenscarl  `num_tokens_across_dp` does an `all_reduce` and it'd be nice if we could avoid it. 
IIUC, the DP padding in `gpu_model_runner` [here](https://github.com/vllm-project/vllm/blob/56bd537dde023f2d8372257255af45fa784ee739/vllm/v1/worker/gpu_model_runner.py#L1510) should make the `local_tokens` across the DP ranks the same. If that is true, then we could simply use `torch.tensor([local_tokens] * get_dp_group().world_size)` in the `all_gatherv` ?

Please let me know if that is not the case. Thanks.

Comment (andoorve) [reply vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py]: @varun-sundar-rabindranath IIUC isn't this only for the non-eager case? Could reuse that logic here.
https://github.com/vllm-project/vllm/blob/56bd537dde023f2d8372257255af45fa784ee739/vllm/v1/worker/gpu_model_runner.py#L1412

Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py]: Yeah the padding happens only for cudagraphs mode. Sorry, didn't realize fix is for eager mode .

Comment (varun-sundar-rabindranath) [reply vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py]: - One  way to avoid the `all_reduce` is to update https://github.com/vllm-project/vllm/blob/56bd537dde023f2d8372257255af45fa784ee739/vllm/v1/worker/gpu_model_runner.py#L1412 such that we pad for this case. 

- another way could be to sneak in the local-sizes via forward-context 🤔  

But I think this is a good fix and I am good with landing this fix as is.

Comment (wenscarl) [reply vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py]: Added a data structure in forward context to keep book of local sizes

**Comment by varun-sundar-rabindranath** [comment]
Hi @wenscarl  - Thanks for implementing this. 

I have some suggestions that I think will make the implementation bit simpler. 
- I believe most of the functionality can be packaged inside DPMetadata which is already a part of the forward context. i.e. if we add 2 things to DPMetadata,
 1. `current_chunk_iteration_idx : Optional[int] = None`
 2. `def get_chunk_sizes_across_dp_rank() -> list[int]`
 
Then in `fused_moe/layer.py` at [here](https://github.com/vllm-project/vllm/blob/97608dc276c292d9217eb6d334d969c5e89913c6/vllm/model_executor/layers/fused_moe/layer.py#L1488), we can do, 
```
get_forward_context().dp_metadata.current_chunk_iteration_idx = chunk_idx
```

and in the all2all [here](https://github.com/vllm-project/vllm/blob/97608dc276c292d9217eb6d334d969c5e89913c6/vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py#L90) we can do,
```
local_sizes = get_forward_context().dp_metadata.get_chunk_sizes_across_dp_rank() 
``` 

I believe this will help silo the DP related changes.

Also, I really like the context manager approach. I think we could preserve it still in the suggested approach. I believe it will simplify things a bit more.

File: vllm/forward_context.py [hunk 3]
```diff
@@ -78,6 +98,30 @@ def make(
...
+        ]
+        self._chunked_local_tokens = _compute_chunked_local_num_tokens(
+            num_tokens_across_dp_cpu, max_chunk_size_per_rank)
+        return
+
+    @contextmanager
+    def chunked_sizes(self, chunk_idx: int):
+        assert self._chunked_local_tokens is not None, (
+            "chunked_local_tokens is not initialized.")
+        self.local_sizes = self._chunked_local_tokens[chunk_idx]
```
Comment (varun-sundar-rabindranath) [inline vllm/forward_context.py]: nit: if `self._chunked_local_tokens[chunk_idx]` is computed on-the-fly here, it will make things more simpler as we could get rid of `set_chunked_local_tokens` (note that this will also make `fused_moe/layer.py`) simpler.

File: vllm/forward_context.py [hunk 4]
```diff
@@ -78,6 +98,30 @@ def make(
...
+            (cu_sizes[i] -
+             cu_sizes[i - 1]).item() if i > 0 else cu_sizes[0].item()
+            for i in range(len(cu_sizes))
+        ]
+        self._chunked_local_tokens = _compute_chunked_local_num_tokens(
+            num_tokens_across_dp_cpu, max_chunk_size_per_rank)
+        return
+
+    @contextmanager
+    def chunked_sizes(self, chunk_idx: int):
```
Comment (varun-sundar-rabindranath) [inline vllm/forward_context.py]: can you add some documentation on why this function is required. Thanks.

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 5]
```diff
@@ -1477,18 +1477,22 @@ def process_chunk(chunk_start, chunk_end, skip_result_store=False):
         max_tokens_across_dp = ctx.dp_metadata.max_tokens_across_dp_cpu
         moe_dp_chunk_size_per_rank = self.moe_config.max_num_tokens
         num_tokens = full_hidden_states.size(0)
-        for chunk_start_ in range(0, max_tokens_across_dp,
-                                  moe_dp_chunk_size_per_rank):
+        ctx.dp_metadata.set_chunked_local_tokens(moe_dp_chunk_size_per_rank)
+        # from vllm.forward_context import with
+        # with set_chunked_tokens_across_dp_cpu(moe_dp_chunk_size_per_rank):
+        for chunk_idx, chunk_start_ in enumerate(
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/fused_moe/layer.py:1548]: remove commented lines before landing.

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 6]
```diff
@@ -1477,18 +1477,22 @@ def process_chunk(chunk_start, chunk_end, skip_result_store=False):
...
             chunk_end = min(chunk_start + moe_dp_chunk_size_per_rank,
                             max_tokens_across_dp)
             # clamp start and end
             chunk_start = min(chunk_start, num_tokens - 1)
             chunk_end = min(chunk_end, num_tokens)
-
-            process_chunk(chunk_start,
-                          chunk_end,
-                          skip_result_store=chunk_start_ >= num_tokens)
+            ctx.dp_metadata.current_chunk_iteration_idx = chunk_idx
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/fused_moe/layer.py]: remove cruft.

File: vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py [hunk 7]
```diff
@@ -90,7 +77,7 @@ def prepare(
             topk_weights, topk_ids, a1q, a1q_scale = \
                 get_dp_group().all_gatherv([topk_weights, topk_ids, a1q, a1q_scale], # noqa: E501
                                            dim=0,
-                                           sizes=get_local_sizes(local_tokens))
+                                           sizes=get_forward_context().dp_metadata.get_chunk_sizes_across_dp_rank()) # noqa: E501
```
Comment (varun-sundar-rabindranath) [inline vllm/model_executor/layers/fused_moe/flashinfer_cutlass_prepare_finalize.py]: nit: maybe get local_sizes in the line above ? so we can avoid  an explicit `#noqa: E501` ?
same comment for the `noqa` line in the finalize function.

**Comment by varun-sundar-rabindranath** [comment]
Hi @wenscarl  Thanks for making the changes. I left some more comments (mostly nits) . This generally looks good to me though. 

cc @bnellnm  @tlrmchlsmth @mgoin

File: vllm/forward_context.py [hunk 8]
```diff
@@ -26,10 +26,29 @@
...
+    for _ in range(chunk_idx + 1):
+        local_sizes = []
+        for i in range(dp_size):
+            to_process = min(remaining[i], max_num_tokens)
+            if to_process == 0:
+                to_process = 1  # ensure lockstep even if done
+            else:
+                remaining[i] -= to_process
+            local_sizes.append(to_process)
+    return local_sizes
```
Comment (tlrmchlsmth) [inline vllm/forward_context.py]: Is this right? We set `local_sizes = []` each iteration inside of the loop but then return it at the end of the function.
Could you double check?

Also instead of using python for loops, it might be better to use numpy operations.

File: vllm/forward_context.py [hunk 9]
```diff
@@ -78,6 +97,48 @@ def make(
         cu_tokens_across_dp_cpu = torch.cumsum(num_tokens_across_dp, dim=0)
         return DPMetadata(max_tokens_across_dp_cpu, cu_tokens_across_dp_cpu)
 
+    @contextmanager
+    def chunked_sizes(self, max_chunk_size_per_rank: int, chunk_idx: int):
```
Comment (tlrmchlsmth) [inline vllm/forward_context.py:98]: Nice - can we use this generally for all fused moe implementations? cc @bnellnm @varun-sundar-rabindranath

File: vllm/forward_context.py (hunk 8 above)
Comment (varun-sundar-rabindranath) [reply vllm/forward_context.py]: I think the function will return the correct answer. It operates primarily on `remaining` tensor to compute the `local_sizes` for each iteration. 
However, I don't think we need do compute the intermediate iteration outputs. We can do something like, 
```
num_chunks = chunk_idx + 1 # I think chunk_idx is 0 based
local_size=[-1] * dp_size
for i in range(dp_size):
      dp_tokens = num_tokens_across_dp_cpu[i]
      local_size[i] = min(max_num_tokens, dp_tokens - (max_num_tokens * num_chunks))
      if local_size[i] <= 0:
          local_size = 1 # ensure lockstep even if done
return local_size
```

File: vllm/forward_context.py (hunk 9 above)
Comment (wenscarl) [reply vllm/forward_context.py:98]: It could be useful if the chunking logic stays the same.

**Review by varun-sundar-rabindranath — APPROVED** [review]
LGTM! @wenscarl  can you fill out the Purpose, Test Plan and Test Result sections of the PR description please. Thanks 🙌

**Comment by wenscarl** [comment]
> LGTM! @wenscarl can you fill out the Purpose, Test Plan and Test Result sections of the PR description please. Thanks 🙌

Done.

**Review by mgoin — APPROVED** [review]
Looks reasonable me, nice work keeping the changes minimal
